### PR TITLE
Add optional wrapping of text for usage messages.

### LINKED ISSUE #41: Help output should be line-wrapped to screen width

### PR DESCRIPTION
**Pull request by gspencergoog**
This adds an _optional_ `maxLineLength` argument to the ArgParser that will automatically wrap help text at the given column on the terminal.

Fixes #41

### CODE REVIEW

**Comment by gspencergoog** [comment]
cc: @nex3 @munificent

File: lib/src/usage.dart [hunk 1]
```diff
@@ -171,8 +176,52 @@ class Usage {
     numHelpLines = 0;
   }
 
+  /// Wrap a single line of text into lines no longer than length.
+  /// Try to split at whitespace, but if that's not good enough to keep it
+  /// under the length, then split in the middle of a word.
+  List<String> wrap(String text, int length) {
```
Comment (munificent) [inline lib/src/usage.dart]: This is only called by `write()`, right?

How about:

* Make it private.
* Move `const int minColumnWidth = 10;` and the width calculation into here instead of passing it in.

File: lib/src/usage.dart [hunk 2]
```diff
@@ -171,8 +176,52 @@ class Usage {
...
+      }
+    }
+    result.add(text.substring(currentLineStart));
+    return result;
+  }
+
   void write(int column, String text) {
+    const int minColumnWidth = 10;
     var lines = text.split('\n');
+    if (column == columnWidths.length && maxLineLength != null) {
```
Comment (munificent) [inline lib/src/usage.dart]: The `column == columnWidths.length` part isn't super obvious. How about adding:

```
// If we are writing the last column, word wrap it to fit.
```

File: lib/src/usage.dart [hunk 3]
```diff
@@ -261,3 +310,20 @@ String padRight(String source, int length) {
 
   return result.toString();
 }
+
+bool isWhitespace(String text, int index) {
+  final int rune = text.codeUnitAt(index);
+  return ((rune >= 0x0009 && rune <= 0x000D) ||
```
Comment (munificent) [inline lib/src/usage.dart]: Nit: None of these parens are needed. If you prefer the ones around the `&&`, that's OK. But the outermost ones are definitely unnecessary.

File: lib/src/usage.dart [hunk 4]
```diff
@@ -171,8 +176,52 @@ class Usage {
...
+    return result;
+  }
+
   void write(int column, String text) {
+    const int minColumnWidth = 10;
     var lines = text.split('\n');
+    if (column == columnWidths.length && maxLineLength != null) {
+      var wrappedLines = <String>[];
+      var start = 0;
+      columnWidths.sublist(0, column).forEach((int width) => start += width);
```
Comment (munificent) [inline lib/src/usage.dart]: I think this is a more idiomatic using `reduce()`. Or just a plain for-in loop.

File: test/usage_test.dart [hunk 5]
```diff
@@ -314,6 +314,102 @@ void main() {
...
+      parser.addFlag('longNewline',
+          help:
+              'The flag with a really long help text and newlines\n\nthat will '
+              'still be wrapped because it is really long.');
+      parser.addFlag('solid',
+          help:
+              'The-flag-with-no-whitespace-that-will-be-wrapped-by-splitting-a-word.');
+      parser.addFlag('small1', help: ' a ');
+      parser.addFlag('small2', help: ' a');
+      parser.addFlag('small3', help: 'a ');
```
Comment (munificent) [inline test/usage_test.dart]: Nice tests!

Is it worth adding a test for how a line containing leading whitespace *and* word-wrapping is handled?

File: lib/src/usage.dart (hunk 1 above)
Comment (gspencergoog) [reply lib/src/usage.dart]: Good suggestions. Done!

File: lib/src/usage.dart (hunk 2 above)
Comment (gspencergoog) [reply lib/src/usage.dart]: Done.

File: lib/src/usage.dart (hunk 4 above)
Comment (gspencergoog) [reply lib/src/usage.dart]: OK.  Switched to reduce().

File: lib/src/usage.dart (hunk 3 above)
Comment (gspencergoog) [reply lib/src/usage.dart]: Removed the parens.

File: test/usage_test.dart (hunk 5 above)
Comment (gspencergoog) [reply test/usage_test.dart]: Sure.  Added, along with one that has trailing whitespace.

**Comment by gspencergoog** [comment]
Superseded by https://github.com/dart-lang/args/pull/107
